src/metabase/driver/sql_jdbc.clj

```diff
@@ -27,11 +27,20 @@
 ;;; |                                                  Run a Query                                                   |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
+(def ^:private ^:dynamic *tx-connections* {})
+
+(defn- local-conn
+  "If called with a transaction for the given database in dynamic scope, return the corresponding connection.
+  Otherwise, return a non-transactional connection."
+  [db-id]
+  (or (get *tx-connections* db-id)
+      (sql-jdbc.conn/db->pooled-connection-spec db-id)))
+
 ;; TODO - Seems like this is only used in a handful of places, consider moving to util namespace
 (defn query
   "Execute a `honeysql-form` query against `database`, `driver`, and optionally `table`."
   ([driver database honeysql-form]
-   (jdbc/query (sql-jdbc.conn/db->pooled-connection-spec database)
+   (jdbc/query (local-conn database)
```

Comment:
We don't yet rely on queries being transactional anywhere, in fact we do not even run any queries within a transaction. This change is done in the spirit of "least surprise" for code written in the future.